server/src/main/java/org/elasticsearch/indices/SystemIndexManager.java

```diff
@@ -0,0 +1,265 @@
+ * under the License.
+ */
+
+package org.elasticsearch.indices;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.mapping.put.PutMappingRequest;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.OriginSettingClient;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.health.ClusterHealthStatus;
+import org.elasticsearch.cluster.health.ClusterIndexHealth;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.MappingMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.routing.IndexRoutingTable;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.gateway.GatewayService;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.cluster.metadata.IndexMetadata.INDEX_FORMAT_SETTING;
+
+/**
+ * This class ensures that all system indices have up-to-date mappings, provided
+ * those indices can be automatically managed. Only some system indices are managed
+ * internally to Elasticsearch - others are created and managed externally, e.g.
+ * Kibana indices.
+ */
+public class SystemIndexManager implements ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(SystemIndexManager.class);
+
+    private final SystemIndices systemIndices;
+    private final Client client;
+    private final AtomicBoolean isUpgradeInProgress;
+
+    public SystemIndexManager(SystemIndices systemIndices, Client client) {
+        this.systemIndices = systemIndices;
+        this.client = client;
+        this.isUpgradeInProgress = new AtomicBoolean(false);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        final ClusterState state = event.state();
+        if (state.blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            // wait until the gateway has recovered from disk, otherwise we may think we don't have some
+            // indices but they may not have been restored from the cluster state on disk
+            logger.debug("system indices manager waiting until state has been recovered");
+            return;
+        }
+
+        // If this node is not a master node, exit.
+        if (state.nodes().isLocalNodeElectedMaster() == false) {
+            return;
+        }
+
```

Comment:
Let's avoid extra work here by guarding this with a `if (upgradeInProgress.get() == false)` check